lib/posix/spawn/child.rb

```diff
@@ -244,14 +251,35 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
 
           # read from stdout and stderr streams
           ready[0].each do |fd|
-            buf = (fd == stdout) ? @out : @err
+            chunk = nil
             begin
-              buf << fd.readpartial(BUFSIZE)
+              chunk = fd.readpartial(BUFSIZE)
             rescue Errno::EAGAIN, Errno::EINTR
             rescue EOFError
               readers.delete(fd)
               fd.close
             end
+
+            abort = false
+            if chunk
+              if fd == stdout
+                if @streaming && @stdout_block
+                  abort = !!@stdout_block.call(chunk)
+                else
+                  @out << chunk
+                end
+              else
+                if @streaming && @stderr_block
+                  abort = !!@stderr_block.call(chunk)
+                else
+                  @err << chunk
+                end
+              end
+            end
+
+            if @streaming && abort
```

Comment:
I think `@streaming &&` is redundant here, since when `@streaming` is not set, `abort` retains its initial value, `false`.